### DISCUSSION TITLE
package conflict problem

### DISCUSSION THREAD

**Discussion by ggslayer** [tauri-apps/tauri#8563] [Q&A] [answered]
Hi, I want to use crate named "systemicons" to fetch file's icon.

but, I occurred a problem like this:

```

error: failed to select a version for `gtk-sys`.
    ... required by package `webkit2gtk v2.0.1`
    ... which satisfies dependency `webkit2gtk = "=2.0.1"` of package `tauri v2.0.0-alpha.20`
    ... which satisfies dependency `tauri = "^2.0.0-alpha.20"` of package `tauri-plugin-global-shortcut v2.0.0-alpha.6`
    ... which satisfies dependency `tauri-plugin-global-shortcut = "^2.0.0-alpha.6"` of package `myappv0.0.0 (F:\myapp\myapp\client\myapp\src-tauri)`
versions that meet the requirements `^0.18.0` are: 0.18.0

the package `gtk-sys` links to the native library `gtk-3`, but it conflicts with a previous package which links to `gtk-3` as well:
package `gtk-sys v0.14.0`
    ... which satisfies dependency `gtk-sys = "^0.14"` of package `systemicons v0.7.0`
    ... which satisfies dependency `systemicons = "^0.7.0"` of package `myapp v0.0.0 (F:\myapp\myapp\client\myapp\src-tauri)`
Only one package in the dependency graph may specify the same links value. This helps ensure that only one copy of a native library is linked in the final binary. Try to adjust your dependencies so that only one package uses the links ='gtk-sys' value. For more information, see https://doc.rust-lang.org/cargo/reference/resolver.html#links.

failed to select a version for `gtk-sys` which could resolve this conflict

```

I don't know how to handle it, can give me some help?

**Comment by FabianLars** [comment, marked as answer]
This can only be really fixed by upgrading gtk inside the systemicons crate. Seeing that it's unmaintained i guess the next best thing is to copy its code into your rust side of your tauri app (using `gtk = "0.18"` in cargo.toml)

**Comment by ggslayer** [comment]
ok Got it



---- Replied Message ----
| From | ***@***.***> |
| Date | 01/07/2024 20:45 |
| To | ***@***.***> |
| Cc | ***@***.***>***@***.***> |
| Subject | Re: [tauri-apps/tauri] package conflict problem (Discussion #8563) |

This can only be really fixed by upgrading gtk inside the systemicons crate. Seeing that it's unmaintained i guess the next best thing is to copy its code into your rust side of your tauri app (using gtk = "0.18" in cargo.toml)

**Reply by S1lander** [reply]
How did you fix this? I tried adding `gtk = "0.18"` to my dependencies, but this didn't solve the error for me.

**Reply by kamiedge** [reply]
Same issue here

**Reply by chenka** [reply]
same issue

**Reply by FabianLars** [reply]
I want to address 2 potential issues here:

1) My comment above was for tauri v2. For tauri v1, the gtk version must be 0.15 (throughout all dependencies)
2) You may missed the important part of my message: 
> Seeing that it's unmaintained i guess the next best thing is to **copy its code into your rust side** of your tauri app

I was basically talking about forking the crate causing you issues so you're able to change its gtk version. In some cases a simpler cargo patch may work but probably not.

**Comment by maxktz** [comment]
added `gtk = "0.18"`

this is what I get:
```
❯ cargo update
    Blocking waiting for file lock on package cache
    Updating crates.io index
error: failed to select a version for `gtk-sys`.
    ... required by package `gtk v0.15.1`
    ... which satisfies dependency `gtk = "^0.15"` of package `tauri v1.6.6`
    ... which satisfies dependency `tauri = "^1.6.5"` of package `app v0.1.0 (/Users/zrxmax/dev/incognet-tauri/src-tauri)`
versions that meet the requirements `^0.15.1` are: 0.15.3, 0.15.1

the package `gtk-sys` links to the native library `gtk-3`, but it conflicts with a previous package which links to `gtk-3` as well:
package `gtk-sys v0.18.0`
    ... which satisfies dependency `ffi = "^0.18"` of package `gtk v0.18.0`
    ... which satisfies dependency `gtk = "^0.18"` of package `app v0.1.0 (/Users/zrxmax/dev/incognet-tauri/src-tauri)`
Only one package in the dependency graph may specify the same links value. This helps ensure that only one copy of a native library is linked in the final binary. Try to adjust your dependencies so that only one package uses the `links = "gtk-3"` value. For more information, see https://doc.rust-lang.org/cargo/reference/resolver.html#links.

failed to select a version for `gtk-sys` which could resolve this conflict
```

**Reply by FabianLars** [reply]
My comment above was for tauri v2. For tauri v1 it must be 0.15 everywhere

**Comment by ansirotenko** [comment]
systemicons 0.8.0 was released recently. Dependency to gtk was updated to 0.18
